### ISSUE TITLE
copy.deepcopy on dunders fails recently

### ISSUE THREAD

**Issue by kiorky** [saltstack/salt#29123]
On our custom modules, we have quite a lot of code that generally clone ``__opts__`` or pillars or grains.

This had worked for years but is falling since 2 weeks.
I think because of https://github.com/saltstack/salt/commit/6838a9534f0dd01291eacfbd6621be43ec0396f4

This is not easily reproducable, sometime it fails, sometimes not ...
But it's driving me crazy

cc @s0undt3ch @jacksontj @thatch45 

```
                      'ContextDict' object has no attribute 'global_data':
                      
                      Traceback (most recent call last):
                        File "/srv/salt/makina-states/mc_states/modules/mc_project_2.py", line 2056, in guarded_step
                          step, cfg['api_version'])](name, *args, **kwargs)
                        File "/srv/salt/makina-states/mc_states/modules/mc_project_2.py", line 2412, in fixperms
                          cret = _step_exec(cfg, 'fixperms')
                        File "/srv/salt/makina-states/mc_states/modules/mc_project_2.py", line 376, in _step_exec
                          cret = _sls_exec(name, cfg, sls)
                        File "/srv/salt/makina-states/mc_states/modules/mc_project_2.py", line 241, in _sls_exec
                          copy.deepcopy(__pillar__))
                        File "/usr/lib/python2.7/copy.py", line 190, in deepcopy
                          y = _reconstruct(x, rv, 1, memo)
                        File "/usr/lib/python2.7/copy.py", line 334, in _reconstruct
                          state = deepcopy(state, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 190, in deepcopy
                          y = _reconstruct(x, rv, 1, memo)
                        File "/usr/lib/python2.7/copy.py", line 334, in _reconstruct
                          state = deepcopy(state, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 190, in deepcopy
                          y = _reconstruct(x, rv, 1, memo)
                        File "/usr/lib/python2.7/copy.py", line 334, in _reconstruct
                          state = deepcopy(state, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
                          y = copier(x, memo)
                        File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
                          y[deepcopy(key, memo)] = deepcopy(value, memo)
                        File "/usr/lib/python2.7/copy.py", line 190, in deepcopy
                          y = _reconstruct(x, rv, 1, memo)
                        File "/usr/lib/python2.7/copy.py", line 358, in _reconstruct
                          y[key] = value
                        File "/salt-venv/salt/src/salt/salt/utils/context.py", line 181, in __setitem__
                          self._dict()[key] = val
                        File "/salt-venv/salt/src/salt/salt/utils/context.py", line 177, in _dict
                          r = r[k]
                        File "/salt-venv/salt/src/salt/salt/utils/context.py", line 111, in __getitem__
                          return self.global_data[key]
                      AttributeError: 'ContextDict' object has no attribute 'global_data'

```

**Comment by kiorky** [comment]
i have a fix.

**Comment by kiorky** [comment]
Well its damn slow... trying better
```
diff --git a/salt/utils/context.py b/salt/utils/context.py
index eefa3a6..c5e3da9 100644
--- a/salt/utils/context.py
+++ b/salt/utils/context.py
@@ -171,6 +171,18 @@ class NamespacedDictWrapper(collections.MutableMapping, dict):
         else:
             self.pre_keys = pre_keys
 
+    def __deepcopy__(self, memo):
+        data = type(self.__dict)()
+        for k, v in salt.ext.six.iteritems(self.__dict):
+            data[k] = copy.deepcopy(v, memo)
+        if isinstance(data, ContextDict):
+            child = data.clone()
+            child.__enter__()
+        pre_keys = copy.deepcopy(self.pre_keys, memo)
+        me = type(self)(data, pre_keys)
+        memo[id(self)] = me
+        return me
+
     def _dict(self):
         r = self.__dict
         for k in self.pre_keys:

```

**Comment by jacksontj** [comment]
Could you include a repro case? It'll make it easier to see what the issue is and if a fix has resolved the problem.

**Comment by kiorky** [comment]
from irc
```
19:46 <kiorky> in this context, copy.deepcopy(__pillar__)
19:46 <kiorky> or
19:46 <kiorky> __opts__ fails
19:46 <kiorky> but i dont know why
19:47 <kiorky> (im exec'in some specific sls in chain, where i inject some other globals, directly in "opts"
19:47 <kiorky> )
19:47 <kiorky> and i restore the opts dict after each separate sls run
19:48 <kiorky> https://github.com/makinacorpus/makina-states/blob/master/mc_states/modules/mc_project_2.py#L240
19:48 <kiorky> this is the line which trigger the error... but not on every project
19:48 <kiorky> and i didnt found why some projects hit the problem, and some others dont
19:50 <jacksontj> why are you injecting things into the pillar?
19:50 <jacksontj> that seems like a mess in and of itself :/
19:51 <kiorky> not in the pillar, in opts.
19:51 <jacksontj> the link you sent to me shows pillar
19:51 <kiorky> this is a specific mecanism to deploy our projects, over salt.
19:51 <kiorky> pillar is copied yes.
19:52 <kiorky> and that you dont understand now why we have something like that is not a reason to tell it's a messs ;)





```

**Comment by kiorky** [comment]
Always trying to find a way to reproduce it in a more canonical way, without sucess so far.

(I repeat the error constantly on one of our projects  through)

**Comment by kiorky** [comment]
i have a fix in our custom code, that seems more a workaround.

So at this point, im not sure it's a salt bug.

**Comment by kiorky** [comment]
(The magic is this line: https://github.com/makinacorpus/makina-states/commit/08d75f68f079ddc78fe3041cdcdad250e616b99b#diff-854538ae720510c8c824c51019fe18efR839
without the deepcopy, the later code, when deepcopying pillar will fail, in certains cases...)
I suspect something deep in the loader, or in my code, but so far i do not have better to tell.

**Comment by cachedout** [comment]
This should have been fixed by #28020, I believe.
